### PR TITLE
gh-116417: Add _testlimitedcapi C extension

### PR DESCRIPTION
**Pull request by vstinner**
Add a new C extension "_testlimitedcapi" which is only built with the limited C API.

Move heaptype_relative.c and vectorcall_limited.c from Modules/_testcapi/ to Modules/_testlimitedcapi/.

* Issue: gh-116417

### CODE REVIEW

**Comment by vstinner** [comment]
@erlend-aasland @encukou @sobolevn @colesbury: What do you think of this change? (See the issue for the rationale.)

File: Modules/_testlimitedcapi/vectorcall_limited.c [hunk 1]
```diff
@@ -2,17 +2,13 @@
 
 #include "pyconfig.h"   // Py_GIL_DISABLED
```
Comment (colesbury) [inline Modules/_testlimitedcapi/vectorcall_limited.c]: I think you can remove this include now that it's part of `parts.h`

**Review by colesbury — COMMENTED** [review]
This seems like a good idea to me.

* The `#include "pyconfig.h"` can be removed from `vectorcall_limited.c` and `heaptype_relative.c` now that it's part of `parts.h`
* `heaptype_relative.c` can also remove the `Py_LIMITED_API` define now that it's defined in `parts.h`.

**Comment by zooba** [comment]
Can we remove `xxlimited.c` then? I'm pretty sure that's only in there to test the limited API, so if we have another module doing it, there isn't a need for that one.

**Comment by vstinner** [comment]
@colesbury:
> This seems like a good idea to me. (...)

Fixed: nicely spotted for remaining code in heaptype_relative.c.

I had a lot of build issues to build the new C extension properly on Linux and Windows, and then I forgot these lines.


@zooba:
> Can we remove xxlimited.c then? I'm pretty sure that's only in there to test the limited API, so if we have another module doing it, there isn't a need for that one.

I don't think that it's a test, but more an example to write code using the limited C API. I started with the comment:

```c
/* Use this file as a template to start implementing a module that
   also declares object types. All occurrences of 'Xxo' should be changed
```

cc @encukou

**Review by colesbury — APPROVED** [review]
LGTM

**Review by corona10 — APPROVED** [review]

**Review by sobolevn — APPROVED** [review]
This looks like a good idea, thank you!

**Review by erlend-aasland — COMMENTED** [review]
I'd like it if you could create an entry for this module in `configure.ac`; that way it is easier to see if this module is enabled or not. Relevant output from sample `./configure` and `make` invocations:

```
The following modules are *disabled* in configure script:
_ctypes_test              _testbuffer               _testcapi              
_testclinic               _testclinic_limited       _testexternalinspection
_testimportmultiple       _testinternalcapi         _testmultiphase        
_testsinglephase          _xxtestfuzz               xxsubtype              
```

```
checking for stdlib extension module _testcapi... disabled
checking for stdlib extension module _testclinic... disabled
checking for stdlib extension module _testclinic_limited... disabled
checking for stdlib extension module _testinternalcapi... disabled
checking for stdlib extension module _testbuffer... disabled
checking for stdlib extension module _testimportmultiple... disabled
checking for stdlib extension module _testmultiphase... disabled
checking for stdlib extension module _testsinglephase... disabled
checking for stdlib extension module _testexternalinspection... disabled
```

If you "hide" this new extension module behind the `_testcapi` module, it will not show up in these outputs.

**Comment by vstinner** [comment]
> I'd like it if you could create an entry for this module in configure.ac; that way it is easier to see if this module is enabled or not.

Sure! It's done.

Example:

```
$ ./configure --disable-test-modules|grep testlimited
checking for stdlib extension module _testlimitedcapi... disabled
```

The test extension is now disabled as expected.

File: Modules/Setup.stdlib.in [hunk 2]
```diff
@@ -162,7 +162,8 @@
...
+@MODULE__TESTCAPI_TRUE@_testcapi _testcapimodule.c _testcapi/vectorcall.c _testcapi/heaptype.c _testcapi/abstract.c _testcapi/bytearray.c _testcapi/bytes.c _testcapi/unicode.c _testcapi/dict.c _testcapi/set.c _testcapi/list.c _testcapi/tuple.c _testcapi/getargs.c _testcapi/datetime.c _testcapi/docstring.c _testcapi/mem.c _testcapi/watchers.c _testcapi/long.c _testcapi/float.c _testcapi/complex.c _testcapi/numbers.c _testcapi/structmember.c _testcapi/exceptions.c _testcapi/code.c _testcapi/buffer.c _testcapi/pyatomic.c _testcapi/pyos.c _testcapi/file.c _testcapi/codec.c _testcapi/immortal.c _testcapi/gc.c _testcapi/sys.c _testcapi/hash.c _testcapi/time.c
+@MODULE__TESTCAPI_TRUE@_testlimitedcapi _testlimitedcapi.c _testlimitedcapi/vectorcall_limited.c _testlimitedcapi/heaptype_relative.c
```
Comment (erlend-aasland) [inline Modules/Setup.stdlib.in]:
```suggestion
@MODULE__TESTLIMITEDCAPI_TRUE@_testlimitedcapi _testlimitedcapi.c _testlimitedcapi/vectorcall_limited.c _testlimitedcapi/heaptype_relative.c
```

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Thanks! On more nit :)

File: Modules/Setup.stdlib.in (hunk 2 above)
Comment (vstinner) [reply Modules/Setup.stdlib.in]: Oh, I don't know well these format. Thanks, I applied your suggestion.

**Comment by erlend-aasland** [comment]
It's a nice change; it makes sense to have this as a separate test module IMO.

**Comment by vstinner** [comment]
Thanks for your helpful reviews! I merged my PR.

**Comment by vstinner** [comment]
Backporting this change to Python 3.12 looks non trivial, I get many conflicts:

```
Unmerged paths:
  (use "git add <file>..." to mark resolution)
	both modified:   Lib/test/support/__init__.py
	both modified:   Lib/test/test_capi/test_misc.py
	both modified:   Modules/Setup.stdlib.in
	both modified:   Modules/_testcapi/parts.h
	both modified:   Modules/_testcapimodule.c
	both modified:   Modules/_testlimitedcapi/heaptype_relative.c
	both modified:   Modules/_testlimitedcapi/vectorcall_limited.c
	both modified:   PCbuild/pcbuild.proj
	both modified:   PCbuild/pcbuild.sln
	both modified:   PCbuild/readme.txt
	both modified:   Tools/c-analyzer/c_parser/preprocessor/gcc.py
	both modified:   Tools/msi/test/test_files.wxs
	both modified:   configure
	both modified:   configure.ac
```

The split of `Modules/_testinternalcapi.c` was not backported. Maybe we should not backport these changes to 3.11 and 3.12.

**Pull request merged**